### PR TITLE
Make `ProofTreeBuilder` actually generic over `Interner`

### PR DESCRIPTION
**Pull request by compiler-errors**
Self-explanatory. Also renamed `ecx.tcx()` to `ecx.interner()`.

r? lcnr

### CODE REVIEW

File: compiler/rustc_middle/src/ty/context.rs [hunk 1]
```diff
@@ -220,6 +220,13 @@ impl<'tcx> Interner for TyCtxt<'tcx> {
         self.check_and_mk_args(def_id, args)
     }
 
+    fn mk_canonical_goal_evaluation_step(
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/context.rs:223]:
```suggestion
    fn intern_canonical_goal_evaluation_step(
```
I feel like `mk` should only be for `XKind -> X`, so this doesn't feel right to me

File: compiler/rustc_trait_selection/src/solve/assembly/mod.rs [hunk 2]
```diff
@@ -455,7 +457,7 @@ impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
         goal: Goal<'tcx, G>,
         candidates: &mut Vec<Candidate<'tcx>>,
     ) {
-        let tcx = self.tcx();
+        let tcx = self.interner();
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/assembly/mod.rs:460]: 🤔 this feels a bit meh, have some thoughts but they aren't blocking so let's chat about this the next time we meet in sync

File: compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs [hunk 3]
```diff
@@ -379,13 +380,17 @@ impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
 /// evaluating a goal. The `var_values` not only include the bound variables
 /// of the query input, but also contain all unconstrained inference vars
 /// created while evaluating this goal.
-pub(in crate::solve) fn make_canonical_state<'tcx, T: TypeFoldable<TyCtxt<'tcx>>>(
-    infcx: &InferCtxt<'tcx>,
-    var_values: &[ty::GenericArg<'tcx>],
+pub(in crate::solve) fn make_canonical_state<
+    Infcx: InferCtxtLike<Interner = I>,
+    I: Interner,
+    T: TypeFoldable<I>,
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs:386]: move to where-clauses please

**Review by lcnr — COMMENTED** [review]
r=me after nits and rebase

File: compiler/rustc_middle/src/ty/context.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/context.rs:223]: We use `mk_` either for `Kind` interning and for some `List`s, but I'd be happy to keep it to just those.

**Comment by workingjubilee** [comment]
conflicts with https://github.com/rust-lang/rust/pull/125343/files

flipping a coin I guess

**Comment by workingjubilee** [comment]
hm PR order should be fine actually.

**Comment by compiler-errors** [comment]
Rebased just in case

@bors r=lcnr

**Pull request merged**